### ISSUE TITLE
Clangd 11 RC1: Extract to function messes up namespaces and strings

### ISSUE THREAD

**Issue by konradsa** [clangd/clangd#524]
Glad to see the extract to function refactoring is getting some love in clangd 11, this would be a powerful code refactoring tool if it would work correctly and reliably. However, I have been playing with it and see some obvious issues.

For example, this is the code before refactoring:

`
#include <iostream>
#include <vector>

namespace foo {

int main() {
  std::vector<std::string> v{"1", "2", "3", "4"};
  std::vector<std::string> v2{"a", "b", "c", "d"};

  for (auto a : v) {
    for (auto a2 : v2) {
      std::cout << a << a2 << std::endl;
    }
  }

  return 0;
}
} // namespace foo
`

Now if I extract the inner loop as a function, I get the following:

`
void extracted(std::std::vector<std::string> &v2,
               basic_string<char, char_traits<char>, allocator<char>> &a) {
  for (auto a2 : v2) {
    std::cout << a << a2 << std::endl;
  }
}
`

Two issues here:
1) Namespace in function signature becomes `std::std::`, which is obviously wrong
2) Instead of `std::string`, it expanded to `basic_string<char, char_traits<char>, allocator<char>>`, which is probably not what most people would want to use here

**Comment by sam-mccall** [comment]
Yep, these are all valid issues. Some notes, for when we pick this up:

> Instead of std::string, it expanded to basic_string<...>

Note that "string" is never spelled in the original code, so it's hard to know (in general) what type is intended. We need some heuristics here, as the AST doesn't contain links from underlying type -> alias.

> Also the basic_string in 2) doesn't compile due to missing namespace specifiers

Right - this is probably as-spelled in std::vector's definition, which doesn't need the qualifiers.

> Both v1 and s1 in extracted could be const qualified, as they are not modified by the function

For v2 this is only half-true: the foreach uses the non-const begin(), and in principle the const and non-const overloads could have different semantics. We need heuristics here again.

> does "Extract to function" only work for free functions and not member functions

Currently, yes.

**Comment by konradsa** [comment]
Thanks for the response Sam. Looks like this feature is still a big work in progress. Would be a very powerful tool though if this could work reliably, including support for member functions.
